utils/training_loop.py

```diff
@@ -1,270 +1,104 @@
+    env = DummyVecEnv([lambda: AirSimGymEnv(config, AirsimManager(config))])
+    model = PPO('MlpPolicy', env, verbose=1, learning_rate=0.0001, n_steps=160, batch_size=160)
+    #model = PPO('MlpPolicy', env, verbose=1)
+    model.set_logger(new_logger)
+
+    def train_model():
+        total_timesteps = 160
+        print('Model learning')
+
+        model.learn(total_timesteps=total_timesteps)
+        print('Model learned ************')
+        env.envs[0].resume_simulation()
 
     if config.ONLY_INFERENCE:
-        rl.network = nn_handler.load_weights_to_network(rl.network)
+        model = PPO.load(config.LOAD_WEIGHT_DIRECTORY)
+        print("Loaded weights for inference.")
 
-    # Initialize counters for the experiment
     collision_counter = 0
     episode_counter = 0
     steps_counter = 0
+    total_steps = 0
 
-    # Main loop for each episode
     for episode in range(config.MAX_EPISODES):
-
-        # choose constant action for car2 for whole trajectory
-        if np.random.randint(2) == 1:
-            config.CAR2_CONSTANT_ACTION = 0
+        print(f"@ Episode {episode + 1} @")
+        car2_side = random.choice(["left", "right"])
+        if car2_side == "left":
+            config.CAR2_INITIAL_POSITION = [0, -30]
+            config.CAR2_INITIAL_YAW = 90
         else:
-            config.CAR2_CONSTANT_ACTION = 1
-
-        episode_counter += 1
-        episode_sum_of_rewards = 0
-        print(f"@ Episode {episode} @")
+            config.CAR2_INITIAL_POSITION = [0, 30]
+            config.CAR2_INITIAL_YAW = 270
 
-        rl.current_trajectory = []
-
-        # Inner loop for each step in an episode
-        for step in range(config.MAX_STEPS):
-            steps_counter += 1
+        print(f"Car 2 starts from {car2_side} with position {config.CAR2_INITIAL_POSITION} and yaw {config.CAR2_INITIAL_YAW}")
 
-            # Perform a step
-            state, action, next_state, collision_occurred, reached_target, reward = rl.step_agent_only()
 
-            # Add current step to trajectory (once for car1 action, and once for car2 (both have same reward))
-            if collision_occurred or reached_target:
-                done = True
-            else:
-                done = False
-            rl.current_trajectory.append((state, action, reward, next_state, done))
+        env.envs[0].airsim_manager.reset_cars_to_initial_positions()
 
-            # if config.TRAIN_OPTION == 'step':
-            #     step_loss = rl.train_trajectory(train_only_last_step=True, episode_counter=episode_counter)
-            #     logger.log_scaler("loss_per_step", steps_counter, step_loss)
+        obs = env.reset()
+        done = False
+        episode_sum_of_rewards = 0
+        episode_counter += 1
 
-            # Update the sum of rewards
+        while not done:
+            # print(f"Before step: Done={done}, State={obs}")
+            if not config.ONLY_INFERENCE:
+                if total_steps > 2500:
```

Comment:
2500 should move to config. (+ add to notion conclusion about how to do exploration/explotation in PPO, I do not know if this is the best way)

Reply:
fixed